coremltools/converters/nnssa/coreml/graph_pass/mlmodel_passes.py

```diff
@@ -244,66 +245,176 @@ def remove_redundant_transposes(spec):
-                    # This layer has one of the deleted as an input. Replace it
-                    # with the deleted layer's input.
-                    idx = [i for i,n in enumerate(_nn_layer.input) if n == _layer_pair[1].name][0]
-                    _nn_layer.input[idx] = _layer_pair[0].input[0]
+        for layers in layers_to_delete:
+            start_layer = layers[0]
+            end_layer = layers[-1]
+
+            # Replace children's input by layer_start's input
+            children = output_to_layers[end_layer.output[0]]
+            for child in children:
+                idx = [i for i, input in enumerate(child.input) if input == end_layer.output[0]]
+                assert(len(idx) == 1)
+                idx = idx[0]
+                child.input[idx] = start_layer.input[0]
+
         # Second pass: delete the layers.
-        for _layer_pair in layers_to_delete:
-            nn_layers.remove(_layer_pair[0])
-            nn_layers.remove(_layer_pair[1])
+        for layers in layers_to_delete:
+            for layer in layers:
+                nn_layers.remove(layer)
 
     def _find_redundant_transposes(nn_spec):
         """
-        Search the neural network spec for pairs of transposes that together
-        are the identity, and return a list of those pairs. This will only find
-        redundant pairs that are adjacent in the topologically sorted layer
-        list and will not find pairs in more complicated structure.
+        Search the neural network spec for sequence of transposes that together
+        are the identity, and return a list of those sequence.
         """
         nn_layers = nn_spec.layers
         layers_to_delete = []
-        # This holds the axes definition if the previous layer was a transpose,
-        # otherwise it is None.
-        previous_transpose = None
-        for _layer in nn_layers:
-            layer_type = _layer.WhichOneof('layer')
-            if layer_type != 'transpose' or len(_layer.output) != 1:
-                previous_transpose = None
-                continue
 
-            if not previous_transpose:
-                previous_transpose = {'layer': _layer, 'axes':_layer.transpose.axes}
-            else:
-                # Check if the layers are connected in the graph.
-                connected = (previous_transpose['layer'].output == _layer.input)
-
-                # Check if they're each other's inverses.
-                if connected:
-                    this_transpose = _layer.transpose.axes
-                    composed = [previous_transpose['axes'][i] for i in this_transpose]
-                    compose_to_identity = all([ax == i for i, ax in enumerate(composed)])
-
-                if connected and compose_to_identity:
-                    # These transpose ops are redundant, remove them.
-                    layers_to_delete.append((previous_transpose['layer'], _layer))
+        # Check no branch or loop in the network...
+        for layer in nn_layers:
```

Comment:
@aseemw  add this check for the branch and loop case.

Reply:
The function doesn't need to return if there is a loop/branch. 
Let it continue. At least it would be able to remove consecutive transposes that are outside of loop/branch body. 

My earlier comment was to just make you aware that the current implementation won't remove tranpose layers from within a loop/branch body, which is fine for now, as it can be handled later.

Reply:
good!